### PR TITLE
Support virtio-transitional for old guests

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

Old guests like RHEL6 or CentOS6 require `virtio-transitional` in newer libvirt/qemu versions. 

Add a `spec.domain.devices.useVirtioTransitional` boolean which indicats which version of `virtio` the virtual machine needs. It will then use `virtio-transitional` for all virtio devices of type:

 * RNG
 * Disks
 * Interfaces
 * virtio-serial controllers

The following stay with their current value:

 * `virtio-scsi` controllers have to stick with `virtio-scsi`
 * `tablet` input devices have to stick with `virtio`
 * Ballooning: due to https://bugzilla.redhat.com/show_bug.cgi?id=1911786

Finally if the `useVirtioTransitional` boolean is unset or false, we now exclititly translate `virtio` from the API to `virtio-non-transitional` to avoid ambiguities, since the term `virtio` is in most cases deprecated now in libvirt and can lead to different results based on different PCI(e) layouts.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

A big chunk of the code changes only extract the domain converter from the domain `api` package into the `converter` package. This was necessary to properly share validation code without import cycles (and improves the code structure).

**Release note**:

```release-note
Add spec.domain.devices.useVirtioTransitional boolean to support virtio-transitional for old guests
```

### CODE REVIEW

**Comment by dankenigsberg** [comment]
/cc @maiqueb @AlonaKaplan 

This seems like a replacement of pull #3510.
The suggested `useVirtioTransitional` API works for me.

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 1]
```diff
@@ -659,10 +663,10 @@ func Convert_v1_Watchdog_To_api_Watchdog(source *v1.Watchdog, watchdog *Watchdog
 	return fmt.Errorf("watchdog %s can't be mapped, no watchdog type specified", source.Name)
 }
 
-func Convert_v1_Rng_To_api_Rng(source *v1.Rng, rng *Rng, _ *ConverterContext) error {
+func Convert_v1_Rng_To_api_Rng(c *ConverterContext, _ *v1.Rng, rng *Rng, _ *ConverterContext) error {
```
Comment (maiqueb) [inline pkg/virt-launcher/virtwrap/api/converter.go]: You already have here a `*ConverterContext` as  a nameless parameter - last parameter in the signature. 

Shouldn't you just rename that one ? 

I'm OK with reordering the parameters, but not w/ duplicating them.

**Review by maiqueb — CHANGES_REQUESTED** [review]
The commit msg heading in https://github.com/kubevirt/kubevirt/pull/4730/commits/bbcb652ff6305b47dfb25fc018e8f6580101d23f has a small typo - s/Table/Tablet/

Other than that, you've duplicated a parameter in the RNG 'converter' function. 
 
Everything else lgtm.

**Comment by maiqueb** [comment]
> /cc @maiqueb @AlonaKaplan
> 
> This seems like a replacement of pull #3510.

It is a replacement for the aforementioned PR.

> The suggested `useVirtioTransitional` API works for me.

The suggested solution is in-line with what was discussed in the aforementioned PR.

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 1 above)
Comment (rmohr) [reply pkg/virt-launcher/virtwrap/api/converter.go]: Fixed. I missed that it was there already.

**Comment by rmohr** [comment]
> The commit msg heading in [bbcb652](https://github.com/kubevirt/kubevirt/commit/bbcb652ff6305b47dfb25fc018e8f6580101d23f) has a small typo - s/Table/Tablet/

Fixed.

**Review by maiqueb — APPROVED** [review]
/lgtm

**Comment by rmohr** [comment]
> I wonder what will happen if you have `useVirtioTransitional` set and install an old operating system such as RHEL 6, though? Since that OS doesn't have virtio 1.0 support, which is the whole reason you'd want to use transitional virtio in the first place, then your disk and network and other devices would work but your pointing device presumably wouldn't? Would it make sense to avoid using virtio-tablet in this scenario, and adding a USB tablet instead?
> 
> >

Yes, one can choose `usb` instead.

> What's the rationale for treating virtio-scsi differently? `<controller type='scsi'>` supports `model='virtio-(non-)transitional'` same as all other devices which are not modern-only, so I would expect `useVirtioTransitional` to affect it too.
> 
> >

The documentation (https://libvirt.org/formatdomain.html#virtio-transitional-devices) said:

```
for SCSI controllers, virtio-scsi must be used instead of virtio for backwards compatibility reasons;
```

I may have misunderstood it.

> I might have missed it, but it doesn't look like there's any check on the machine type being performed. While it's true that for q35 machines virtio and virtio-non-transitional will behave the same, for pc machines virtio will behave like virtio-transitional instead, and I'm afraid explicitly choosing virtio-non-transitional will result in broken pc machines. But perhaps that's not considered a concern for KubeVirt?

I did not know that. I probably missed it in the documentation. We only support `q35`. However, one can whitelist other machine types, but we never test them. I guess as a workaround one would have to set `useVirtionTransitional` with these types. I would say that it is at least not optional.

**Comment by andreabolognani** [comment]
> > I wonder what will happen if you have `useVirtioTransitional` set and install an old operating system such as RHEL 6, though? Since that OS doesn't have virtio 1.0 support, which is the whole reason you'd want to use transitional virtio in the first place, then your disk and network and other devices would work but your pointing device presumably wouldn't? Would it make sense to avoid using virtio-tablet in this scenario, and adding a USB tablet instead?
> 
> Yes, one can choose `usb` instead.

Does that have to happen explicitly, though? My point is that setting `useVirtioTransitional` should probably do that already, since by setting it the user has already made it clear that they don't want modern-only devices, and virtio-tablet is modern-only.

> > What's the rationale for treating virtio-scsi differently? `<controller type='scsi'>` supports `model='virtio-(non-)transitional'` same as all other devices which are not modern-only, so I would expect `useVirtioTransitional` to affect it too.
> 
> The documentation (https://libvirt.org/formatdomain.html#virtio-transitional-devices) said:
> 
> ```
> for SCSI controllers, virtio-scsi must be used instead of virtio for backwards compatibility reasons;
> ```
> 
> I may have misunderstood it. It probably just meant that it never had `virtio` as an option.

Yeah, the documentation is not very clear on the topic. I've just pushed a commit that improves upon it.

https://gitlab.com/libvirt/libvirt/-/commit/85523cfae0be52868df26463650dd8e40a156fb9

The change should be reflected to the website in a while.

> > I might have missed it, but it doesn't look like there's any check on the machine type being performed. While it's true that for q35 machines virtio and virtio-non-transitional will behave the same, for pc machines virtio will behave like virtio-transitional instead, and I'm afraid explicitly choosing virtio-non-transitional will result in broken pc machines. But perhaps that's not considered a concern for KubeVirt?
> 
> I did not know that. I probably missed it in the documentation. We only support `q35`. However, one can whitelist other machine types, but we never test them. I guess as a workaround one would have to set `useVirtionTransitional` with these types. I would say that it is at least not optimal.

I'm not sure what the best course of action would be for KubeVirt, but I wanted to make sure you were aware of the potential issues with this implementation.

**Comment by rmohr** [comment]
@andreabolognani I will prepare a follow-up PR. Thanks for all the details.

**Comment by andreabolognani** [comment]
@rmohr great!

What about the fix for [RHBZ#1911786](https://bugzilla.redhat.com/show_bug.cgi?id=1911786)? Should I try to prepare a backport and update the virt-launcher container image accordingly? That'd force me to get at least somewhat familiar with the new bazel-based image building process O:-)

**Comment by rmohr** [comment]
> What about the fix for [RHBZ#1911786](https://bugzilla.redhat.com/show_bug.cgi?id=1911786)? Should I try to prepare a backport and update the virt-launcher container image accordingly?

A backport would be great. The more we can do to ensure the RHEL6 and CentOS6 is happy, the better.

> That'd force me to get at least somewhat familiar with the new bazel-based image building process O:-)

Looking forward to your honest feedback ;)

**Comment by maya-r** [comment]
I am under the impression that transitional virtio supports both virtio<1.0 and virtio>=1.0 
Have we considered not having a tunable for it and making it the default?

**Comment by rmohr** [comment]
> I am under the impression that transitional virtio supports both virtio<1.0 and virtio>=1.0
> Have we considered not having a tunable for it and making it the default?

virtio-transitional is the implementation of the virtio 0.9x specification. virtio-non-transitional is virtio 1.0. We basically present different hardware to the guest.

**Comment by maya-r** [comment]
virtio transitional is magically both 0.9 and >=1.0.
Virtio-1.0 capable drivers know to read VIRTIO_F_VERSION_1 and tell that >=1.0 is supported.

**Comment by rmohr** [comment]
> virtio transitional is magically both 0.9 and >=1.0.
> Virtio-1.0 capable drivers know to read VIRTIO_F_VERSION_1 and tell that >=1.0 is supported.

You are right, but still:

```
virtio-transitional
This device can work both with virtio 0.9 and virtio 1.0 guest drivers, so it's the best choice when compatibility with older guest operating systems is desired. libvirt will plug the device into a conventional PCI slot.

virtio-non-transitional
This device can only work with virtio 1.0 guest drivers, and it's the recommended option unless compatibility with older guest operating systems is necessary. libvirt will plug the device into either a PCI Express slot or a conventional PCI slot based on the machine type, resulting in a more optimized PCI topology.
```

if we don't want to manage the PCI topology completely ourselves, we should choose based on the OS.

**Pull request merged**